File: Searching-Array/prg.dart
import 'dart:io';

void linearSearch(List<int> arr, int key) {
  int index = arr.indexOf(key);
  if (index != -1) {
    print("Found at index $index");
  } else {
    print("Not found");
  }
}

void bubbleSort(List<int> arr) {
  for (int i = 0; i < arr.length - 1; i++) {
    for (int j = 0; j < arr.length - i - 1; j++) {
      if (arr[j] > arr[j + 1]) {
        int temp = arr[j];
        arr[j] = arr[j + 1];
        arr[j + 1] = temp;
      }
    }
  }
  print("Sorted Array: $arr");
}

int interpolationSearch(List<int> arr, int key) {
  int low = 0, high = arr.length - 1;
  while (low <= high && key >= arr[low] && key <= arr[high]) {
    int pos = low + ((key - arr[low]) * (high - low) ~/ (arr[high] - arr[low]));
    if (arr[pos] == key) return pos;
    if (arr[pos] < key) {
      low = pos + 1;
    } else {
      high = pos - 1;
    }
  }
  return -1;
}

void main() {
  List<int> arr = [10, 20, 30, 40, 50];
  while (true) {
    print("1. Linear Search\n2. Bubble Sort\n3. Interpolation Search\n4. Exit");
    int? choice = int.tryParse(stdin.readLineSync()!);
    if (choice == null || choice == 4) break;
    if (choice == 1 || choice == 3) {
      print("Enter key to search: ");
      int? key = int.tryParse(stdin.readLineSync()!);
      if (key == null) continue;
      if (choice == 1) {
        linearSearch(arr, key);
      } else {
        print("Found at index: ${interpolationSearch(arr, key)}");
      }
    } else if (choice == 2) {
      bubbleSort(arr);
    }
  }
}

File: Searching-Array/prg.scala
import scala.io.StdIn._

object SearchSortMenu {
  def linearSearch(arr: Array[Int], key: Int): Unit = {
    val index = arr.indexOf(key)
    if (index != -1) println(s"Found at index $index")
    else println("Not found")
  }

  def bubbleSort(arr: Array[Int]): Unit = {
    for (i <- arr.indices; j <- 0 until arr.length - i - 1)
      if (arr(j) > arr(j + 1)) {
        val temp = arr(j)
        arr(j) = arr(j + 1)
        arr(j + 1) = temp
      }
    println("Sorted Array: " + arr.mkString(", "))
  }

  def interpolationSearch(arr: Array[Int], key: Int): Int = {
    var low = 0
    var high = arr.length - 1
    while (low <= high && key >= arr(low) && key <= arr(high)) {
      val pos = low + ((key - arr(low)) * (high - low) / (arr(high) - arr(low)))
      if (arr(pos) == key) return pos
      if (arr(pos) < key) low = pos + 1
      else high = pos - 1
    }
    -1
  }

  def main(args: Array[String]): Unit = {
    val arr = Array(10, 20, 30, 40, 50)
    while (true) {
      println("1. Linear Search\n2. Bubble Sort\n3. Interpolation Search\n4. Exit")
      val choice = readInt()
      if (choice == 4) return
      if (choice == 1 || choice == 3) {
        print("Enter key to search: ")
        val key = readInt()
        if (choice == 1) linearSearch(arr, key)
        else println("Found at index: " + interpolationSearch(arr, key))
      } else if (choice == 2) {
        bubbleSort(arr)
      }
    }
  }
}
